# 체크 예외와 인터페이스
![](https://i.imgur.com/cpQm1QI.png)
- 서비스 계층은 특정 구현기술에 의존하지 않고, 순수하게 유지하는 것이 좋다. 
    - 이렇게 하려면 예외에 대한 의존도 함께 해결해야한다.
        - ex) SQLException 에 대한 의존을 제거하려면
            - 서비스가 처리할 수 없으므로 리포지토리가 던지는 SQLException 체크 예외를 
            - 런타임 예외로 전환해서 서비스 계층에 던진다
## 인터페이스 도입
```java
public interface MemberRepository {
    Member save(Member member);
    Member findById(String memberId);
    void update(String memberId, int money);
    void delete(String memberId);
}
```
- 특정 기술에 종속되지 않는 순수한 인터페이스이다. 
    - 이 인터페이스를 기반으로 특정 기술을 사용하는 구현체를 만들면 된다.
- 그런데 체크 예외를 사용하려면 인터페이스에도 해당 체크 예외가 선언 되어 있어야 한다.
```java
인터페이스>>

public interface MemberRepositoryEx {
    Member save(Member member) throws SQLException;
    Member findById(String memberId) throws SQLException;
    void update(String memberId, int money) throws SQLException;
    void delete(String memberId) throws SQLException;
}

---
구현클래스>>

public class MemberRepositoryV3 implements MemberRepositoryEx {
    public Member save(Member member) throws SQLException {
        String sql = "insert into member(member_id, money) values(?, ?)";
    } 
}
```
- 인터페이스의 구현체가 체크 예외를 던지려면, 인터페이스 메서드에 먼저 체크 예외를 던지는 부분이 선언 되어 있어야 한다.     - 그래야 구현 클래스의 메서드도 체크 예외를 던질 수 있다.
- 참고로 구현 클래스의 메서드에 선언할 수 있는 예외는 부모 타입에서 던진 예외와 같거나 하위 타입이어야 한다.
    - 인터페이스 메서드에 throws Exception 를 선언하면, 구현 클래스 메서드에 throws SQLException 는 가능하다. SQLException 은 Exception 의 하위 타입이기 때문이다.
- 체크예외의 인터페이스는 특정 기술에 종속하게 된다
    - 인터페이스를 만드는 목적은 구현체를 쉽게 변경하기 위함인데, 이미 **인터페이스가 특정 구현 기술에 오염이 되어 버렸다**. 
    - 향후 JDBC가 아닌 **다른 기술로 변경한다면 인터페이스 자체를 변경해야 한다**.

# 런타임과 인터페이스 
- 예외 누수 문제 해결
- 체크 예외를 런타임 예외로 변경
    - throws SQLException 제거
- MemberRepository 인터페이스 사용 

```java
public class MyDbException extends RuntimeException {
    public MyDbException() {
    }
    public MyDbException(String message) {
        super(message);
    }
    public MyDbException(String message, Throwable cause) {
        super(message, cause);
    }
    public MyDbException(Throwable cause) {
          super(cause);
    }
}
---
구현 클래스>>

public Member save(Member member) {
    try {
        //비즈니스 로직
        } catch (SQLException e) {
            throw new MyDbException(e);
        } finally {
            ..
        }
}

```
1. RuntimeException 을 상속받았다. 따라서 MyDbException 은 런타임(언체크) 예외가 된다.
2. 기존 예외를 생성자를 통해서 포함해야 
    - 예외는 원인이 되는 예외를 내부에 포함할 수 있는데, 꼭 이렇게 작성해야 한다. 
    - 그래야 예외를 출력했을 때 원인이 되는 기존 예외도 함께 확인할 수 있다.
    - MyDbException 이 내부에 SQLException 을 포함하고 있다고 이해하면 된다. 
        - 예외를 출력했을 때 스택 트레이스를 통해 둘다 확인할 수 있다.
    - 원인이 되는 예외를 내부에 포함하지 않으면, 예외를 스택 트레이스를 통해 출력했을 때 기존에 원인이 되는 부분을 확인할 수 없다.
        - 만약 SQLException 에서 문법 오류가 발생했다면 그 부분을 확인할 방법이 없게 된다.

# 데이터 접근 예외 직접 만들기
- 특정 상황에는 예외를 잡아서 복구하고 싶으면 예외를 어떻게 구분해서 처리할 수 있을까?
- 데이터베이스 오류에 따라서 특정 예외는 복구하고 싶을 수 있다.
    - 회원 가입시 DB에 이미 같은 ID가 있으면 ID 뒤에 숫자를 붙여서 새로운 ID를 만들어야 한다고 가정
    - ID를 hello 라고 가입 시도 했는데, 이미 같은 아이디가 있으면 hello12345 와 같이 뒤에 임의의 숫자를 붙여서 가입하는 것

![](https://i.imgur.com/0z5XD9x.png)

**SQL ErrorCode**
- SQLException 내부에 들어있는 errorCode 를 활용하면 데이터베이스에서 어떤 문제가 발생했는지 확인할 수 있다.
    - e.getErrorCode() == 23505
    - 참고로 같은 오류여도 각각의 데이터베이스마다 정의된 오류 코드가 다르다. 따라서 오류 코드를 사용할 때는 데이터베이스 메뉴얼을 확인해야 한다.

**예외 변환**
1. 서비스 계층
    - 예외 복구를 위해 키 중복 오류를 확인
    - 새로운 ID를 만들어서 다시 저장을 시도
2. 리포지토리
    - SQLException 을 서비스 계층에 던지고 서비스 계층은 이 예외의 오류 코드를 확인해서 키 중복 오류( 23505 )인 경우 새로운 ID를 만들어서 다시 저장
3. 특정 기술에 의존해야하는 문제
    - 그런데 이렇게 SQLException 에 들어있는 오류 코드를 활용하기 위해 SQLException 을 서비스 계층으로 던지게 되면, 서비스 계층이 SQLException 이라는 JDBC 기술에 의존하게 됨
    - 지금까지 우리가 고민했던 서비스 계층의 순수성이 무너진다.
4. 이 문제를 해결하려면 앞서 배운 것 처럼 **리포지토리에서 예외를 변환**해서 던지면 된다. 
    - SQLException -> MyDuplicateKeyException
    - MyDuplicateKeyException 예외는 데이터 중복의 경우에만 던져야  한다.
    - 이 예외는 우리가 직접 만든 것이기에 특정 기술에 종속적이지 않다. 
    - 따라서 이 예외를 사용하더라도 서비스 계층의 순수성을 유지할 수 있다. 
    - (향후 JDBC에서 다른 기술로 바꾸어도 이 예외는 그대로 유지할 수 있다.)
5. 하지만 여전히 문제점은 있다
    - SQL ErrorCode는 각각의 데이터베이스 마다 다르다. 
        - 데이터베이스가 변경될 때 마다 ErrorCode도 모두 변경해야 한다.
    - 데이터베이스가 전달하는 오류는 수십 수백가지 오류 코드가 있다.    
        - 이 모든 상황에 맞는 예외를 지금처럼 다 만들어야 할까? 추가로 앞서 이야기한 것 처럼 데이터베이스마다 이 오류 코드는 모두 다르다.
```java
서비스>>

try {
    //비즈니스 로직
} catch (MyDuplicateKeyException e) {
    log.info("키 중복, 복구 시도");
    String retryId = generateNewId(memberId); 
    log.info("retryId={}", retryId); 
    repository.save(new Member(retryId, 0));
} catch (MyDbException e) { 
    log.info("데이터 접근 계층 예외", e);
    throw e; 
}
---
리포지토리>>

catch (SQLException e) {
    //h2 db
    if (e.getErrorCode() == 23505) {
    throw new MyDuplicateKeyException(e);
    }
    throw new MyDbException(e);
}
```
- 리포지토리
    1. e.getErrorCode() == 23505 : 오류 코드가 키 중복 오류( 23505 )인 경우 MyDuplicateKeyException 을 새로 만들어서 서비스 계층에 던진다. 
    2. 나머지 경우 기존에 만들었던 MyDbException 을 던진다.
- 서비스
    1. 저장을 시도
    2. 만약 리포지토리에서 MyDuplicateKeyException 예외가 올라오면 이 예외를 잡는다.
        - 예외를 잡아서 generateNewId(memberId) 로 새로운 ID 생성을 시도한다. 그리고 다시 저장한다.
        - 예외를 복구하는 부분
    3. 만약 복구할 수 없는 예외( MyDbException )면 로그만 남기고 다시 예외를 던진다.
    - 참고로 이 경우 여기서 예외 로그를 남기지 않아도 된다. 
        - 어차피 복구할 수 없는 예외는 예외를 공통으로 처리하는 부분까지 전달되기 때문
        - 사실, 이렇게 복구 할 수 없는 예외는 공통으로 예외를 처리하는 곳에서 예외 로그를 남기는 것이 좋다. 
        - 지금은 예외를 다양하게 잡아서 처리가능하다는 것을 보여주기 위한 예시
# 스프링 예외 추상화